docs/shared_library/scryer_python.py

```diff
@@ -0,0 +1,426 @@
+def query_generator(machine: MACHINE_PTR, query: str) -> ContextManager[Generator[SCRYER_QUERY_RESULT, None, None]]:
+    """
+    Generates a query generator for a given machine and query.
+
+    :param machine: A pointer to the machine.
+    :param query: The query string.
+    :return: A generator that yields the results of running the generator step.
+    """
+    qs: QUERY_STATE_PTR = scryer_run_query_iter(machine, query.encode('utf-8'))
+    deloaded = False
+    try:
+        def g():
+            nonlocal deloaded
+            while True:
+                answer = run_generator_step(qs)
+                if answer is None:
+                    handle_loader_deloader(scryer_query_state_free(qs))
+                    deloaded = True
+                    break
+                yield answer[0] if isinstance(answer, list) else answer
+
+        yield g()
+    finally:
+        if deloaded is False:
+            handle_loader_deloader(scryer_query_state_free(qs))
+
+
+def consult(machine: MACHINE_PTR, facts: str, module_name: str = None):
+    """
+    Consults the given facts in the specified machine. Effectively the same as `load_facts()`.
+
+    :param module_name: name of module to install facts, optional
+    :param machine: Pointer to the machine object.
+    :type machine: MACHINE_PTR
+    :param facts: The facts to be consulted.
+    :type facts: str
+    :return: The result of consulting the facts.
+    """
+    if module_name and isinstance(module_name, str):
+        facts, module_name = module_name, facts
+    else:
+        module_name = 'facts'
+    handle_loader_deloader(scryer_consult_module_string(machine, module_name.encode('UTF-8'), facts.encode('utf-8')))
+
+
+####################
+## SCRYER MACHINE ##
+####################
+
+## This is a suggested high level API that prevents most footguns as long as you stay within the context manager API.
+## see examples below for intended usage and example queries.
+## For more information on using Scryer Prolog, check out https://www.metalevel.at/prolog and
+## https://www.youtube.com/@ThePowerOfProlog
+class ScryerMachine:
+    class IllegalWAMStateError(RuntimeError):
+        pass
+
+    def __init__(self, source=None):
+        """
+        Constructor for the class.
+
+        :param source: The source of the object.
+        :type source: Any, optional
+        """
+        self.source = source
+        self.m: MACHINE_PTR = None
+        self.query_in_progress = False
+
+    def __enter__(self, query: str = None):
+        """
+        Enter method for context manager.
+
+        :param query: Optional query string to be evaluated.
+        :return: If query is provided, return the lazy evaluation of the query using the scryer machine.
+                 Otherwise, return self.
+        """
+        self.m = scryer_machine_new()
+        if self.source:
+            self.consult(self.source)
+        if query:
+            return self.lazy_eval_context(query)
```

Comment:
Giving this a bit of a test drive since it's perfect for a prototype I am working on, so thanks so much! 

I just wanted to point out quickly that returning two different things like this causes some Python typing issues when used in the way of the examples:

![image](https://github.com/user-attachments/assets/4faa1513-a114-4b0a-a3aa-d69fb1aed2d7)



```
Cannot access attribute "lazy_eval_context" for class "ContextManager[Generator[SCRYER_QUERY_RESULT, None, None]]"
  Attribute "lazy_eval_context" is unknown (Pyright reportAttributeAccessIssue)
──────────────────────────────────────────────────────────────────────────────
https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportAttributeAccessIssue


```